[ashell] Fixing duplicate characters on UART

brianjjones opened the pull request:
This fixes the duplicate character issue on UART, removes a bunch
of unused strings, and functions that were redundant.

Signed-off-by: Brian J Jones <brian.j.jones@intel.com>

pfalcon commented on src/ashell/comms-uart.c:
```diff
@@ -322,31 +303,18 @@ void comms_write_buf(const char *buf, int len)
     data_transmitted = false;
-    uart_fifo_fill(dev, buf, len);
-
+    uart_fifo_fill(dev_upload, (const uint8_t *)buf, len);
```
FYI, this is not a correct way to use this function. It returns number of bytes actually written, which can be less than requested, including 0: https://github.com/zephyrproject-rtos/zephyr/blob/cc334c72732bbb23d7809f5f010ecb32345afd62/include/uart.h#L248

brianjjones replied:
This is how all the Zephyr examples use it, my understanding is it means I don't want to do anything specific if the uart fails to copy my entire buffer.  I hate to block here and keep trying to get the uart_fifo_fill to succeed, as perhaps it could go on forever.  I'm open to suggestions however, is it expected we would keep trying until we are sure we had sent all the bytes? At what point do we give up? Perhaps @grgustaf has some insight?

Thanks for the comment.

grgustaf replied:
You're already busy waiting on data_transmitted so if that's anywhere close to okay, you should be fine to expand the loop to retry on the remaining bytes, until all your bytes are gone.

So basically:
```
while (len) {
    len -= uart_blah()
}

would maybe do it? Let's hear from our resident UART expert @jprestwo. :)

grgustaf replied:
I see you're right, some samples do exactly what you did, others ignore the return value just assume it works, without even the data transmitted loop.

The function @pfalcon pointed to shows that it will return 0 if there is no such api in the driver; so if that could ever happen we would hang in my above code. Maybe then 0 should be considered an error? Not sure, the samples and documentation are insufficient.

`test_uart_fifo.c` shows this in `test_fifo_fill`:
```c
		data_transmitted = false;
		while (!uart_fifo_fill(uart_dev, (uint8_t *) &fifo_data[i], 1));
		while (data_transmitted == false);
```

so I guess they're expecting 0 to be a recoverable error.

The only other sample code that ever checks the return value of the function is in `drivers/bluetooth/hci/h4.c` which considers it an error if it's not 1 (they were trying to write just one byte).

Maybe we should consider 0 an error for now to avoid hanging? Otherwise there doesn't seem to be any way to distinguish "the buffer's just full, I can't write any bytes" from "something's not working and will never work".

pfalcon replied:
> The function @pfalcon pointed to shows that it will return 0 if there is no such api in the driver; so if that could ever happen we would hang in my above code.

Well, that's not the problem to care about. If CONFIG_UART_INTERRUPT_DRIVEN is defined, that function can't be non-empty, it's part of the core interrupt-driven API. The problem is like above - you can't expect it to write everything you asked it for.

Code like that (not checking return) is the underlying cause of all those "can't get lines longer 90 chars" and "it doesn't work on any other hardware" problems.

grgustaf commented on src/ashell/jerry-code.c:
```diff
@@ -356,28 +356,6 @@ int javascript_parse_code(const char *file_name, bool show_lines)
-            }
-            comms_println("[END]");
-        }
-
```
Are you taking this out to save space? Seems kind of neat but I didn't know about it. You could give a 'v' arg on the command line when displaying code and get line numbers?

brianjjones replied:
Yes its basically redundant code, all its doing is printing the file with lines.  If someone REALLY wants to print the file with lines I can add that to the ashell_print_file command, that's where it belongs.  But again I don't see the point personally, the IDE has line numbers and is far easier to read.  And size is a real problem now.

grgustaf commented on src/ashell/comms-uart.c:
```diff
@@ -322,31 +303,18 @@ void comms_write_buf(const char *buf, int len)
 }
-
-void comms_println(const char *buf)
+void comms_print(const char *buf, bool newline)
```
If you do insist on removing writec, then it's cleaner to use

 ```c
write_buf(" ", 1)
```
than
```c
char space = ' ';
write_buf(&space, 1);
```

brianjjones replied:
Agreed, that was leftover from testing.  And yes I don't want to have several comms_write_* functions, it makes it confusing (to me at least) what exactly is being done in each.

brianjjones replied:
Looking at it, it seems to make more sense to remove the newline arg and just add them to the msg strings.  I still get to remove the extra function, and its clearer what is being done.

brianjjones commented on src/ashell/shell-state.c:
```diff
@@ -282,37 +272,38 @@ int32_t ashell_print_file(char *buf)
 
+    int lineStart = 0;
     do {
-        count = fs_read(file, data, 4);
```
Rather than reading and printing one character at a time, it does up to 80.  Note comms_printf is no longer used for critical output as it will occasionally add extra characters.

brianjjones commented on src/ashell/shell-state.c:
```diff
@@ -644,12 +618,14 @@ static const struct ashell_cmd commands[] =
+    comms_print("Commands list:\r\n");
     for (uint32_t t = 0; t < ASHELL_COMMANDS_COUNT; t++) {
-        comms_printf("%8s %s\r\n", commands[t].cmd_name, commands[t].syntax);
+        comms_print(commands[t].cmd_name);
```
Again using comms_print and comms_write_buf instaed of comms_printf to avoid extra chars.

brianjjones commented:
To make this clear, comms_printf and comms_writec appear to be the culprits in adding extra characters.  I removed comms_writec and now use comms_write_buf exclusively.  I also removed some functions I felt were redundant.  Sorry I didn't break those into another patch, was on a cleanup mission and didn't think about it until I had put it all together.  Let me know if anything else still isn't clear.

grgustaf commented on src/ashell/comms-uart.c:
```diff
@@ -303,7 +284,7 @@ static void comms_interrupt_handler(struct device *dev)
 static int comms_out(int c)
 {
-    comms_writec((char)c);
+    comms_write_buf((char*)((int*)&c), 1);
```
Could you change that to this:

```
char ch = (char)c;
comms_write_buf(&c, 1);
```

grgustaf commented:
+1, I verified this helps the problem a lot. But that's on output (like `cat [filename]`) not on input echoing (like `load [filename]` + paste). So more to be done!
